rest/stargate.go

```diff
@@ -0,0 +1,129 @@
+package rest
+
+import (
+	"fmt"
+	"net/url"
+	"strings"
+
+	"github.com/cosmos/cosmos-sdk/types/query"
+	"github.com/cosmos/cosmos-sdk/types/tx"
+	"github.com/gin-gonic/gin"
+	"github.com/jackc/pgx/v4/pgxpool"
+	"github.com/likecoin/likecoin-chain-tx-indexer/db"
+	"github.com/likecoin/likecoin-chain-tx-indexer/logger"
+)
+
+func getOffset(query url.Values) (uint64, error) {
+	return getUint(query, "pagination.offset")
+}
+
+func isOrderByDesc(query url.Values) (bool, error) {
+	orderByStr := strings.ToUpper(query.Get("order_by"))
+	switch orderByStr {
+	case "", "ORDER_BY_UNSPECIFIED", "ORDER_BY_ASC":
+		return false, nil
+	case "ORDER_BY_DESC":
+		return true, nil
+	default:
+		return false, fmt.Errorf("available values for order_by: ORDER_BY_UNSPECIFIED, ORDER_BY_ASC, ORDER_BY_DESC")
+	}
+}
+
+func trimSingleQuotes(s string) (string, error) {
+	if len(s) >= 2 {
+		if c := s[len(s)-1]; s[0] == c && c == '\'' {
+			return s[1 : len(s)-1], nil
+		}
+		return "", fmt.Errorf("expect query event value missing single quotes: %s", s)
+	}
+	return "", fmt.Errorf("invalid query event value: %s", s)
+}
+
+func getEventMap(eventArray []string) (url.Values, error) {
+	m := make(url.Values)
+	for _, v := range eventArray {
+		if strings.Contains(v, "=") {
+			arr := strings.SplitN(v, "=", 2)
+			value, err := trimSingleQuotes(arr[1])
+			if err != nil {
+				return nil, err
+			}
+			m[arr[0]] = []string{value}
+		} else {
+			return nil, fmt.Errorf("query event missing equal sign: %s", v)
+		}
+	}
+	return m, nil
+}
+
+func handleStargateTxsSearch(c *gin.Context, pool *pgxpool.Pool) {
+	q := c.Request.URL.Query()
+
+	offset, err := getOffset(q)
+	if err != nil {
+		c.AbortWithStatusJSON(400, gin.H{"error": err.Error()})
+		return
+	}
+	limit, err := getLimit(q, "pagination.limit")
+	if err != nil {
+		c.AbortWithStatusJSON(400, gin.H{"error": err.Error()})
+		return
+	}
+	offsetInTimesOfLimit := offset / limit * limit // Cosmos' bug? 29 / 10 * 10 = 20
```

Comment:
So the Cosmos API actually round the `offset` to a multiple of `limit`?

Reply:
Yes, for example, if `limit = 10`, there is no diff for `offset = 20` and `offset = 29`.